mlkem/native/aarch64/src/arith_native_aarch64.h

```diff
@@ -31,23 +31,53 @@ extern const int16_t mlk_aarch64_zetas_mulcache_twisted_native[];
 extern const uint8_t mlk_rej_uniform_table[];
 
 #define mlk_ntt_asm MLK_NAMESPACE(ntt_asm)
-void mlk_ntt_asm(int16_t *, const int16_t *, const int16_t *);
+void mlk_ntt_asm(int16_t *p, const int16_t *, const int16_t *)
+/* This must be kept in sync with the HOL-Light specification
+ * in proofs/hol_light/arm/proofs/mlkem_ntt.ml */
+__contract__(
+  requires(memory_no_alias(p, sizeof(int16_t) * MLKEM_N))
+  requires(array_abs_bound(p, 0, MLKEM_N, 8192))
+  assigns(memory_slice(p, sizeof(int16_t) * MLKEM_N))
+  ensures(array_abs_bound(p, 0, MLKEM_N, 23595))
+);
 
 #define mlk_intt_asm MLK_NAMESPACE(intt_asm)
-void mlk_intt_asm(int16_t *, const int16_t *, const int16_t *);
+void mlk_intt_asm(int16_t *p, const int16_t *, const int16_t *)
+/* This must be kept in sync with the HOL-Light specification
+ * in proofs/hol_light/arm/proofs/mlkem_intt.ml */
+__contract__(
+  requires(memory_no_alias(p, sizeof(int16_t) * MLKEM_N))
+  assigns(memory_slice(p, sizeof(int16_t) * MLKEM_N))
+  ensures(array_abs_bound(p, 0, MLKEM_N, 26625))
+);
 
 #define mlk_poly_reduce_asm MLK_NAMESPACE(poly_reduce_asm)
 void mlk_poly_reduce_asm(int16_t *);
 
 #define mlk_poly_tomont_asm MLK_NAMESPACE(poly_tomont_asm)
-void mlk_poly_tomont_asm(int16_t *);
+void mlk_poly_tomont_asm(int16_t *p)
+/* This must be kept in sync with the HOL-Light specification
+ * in proofs/hol_light/arm/proofs/mlkem_poly_tomont.ml */
+__contract__(
+  requires(memory_no_alias(p, sizeof(int16_t) * MLKEM_N))
+  assigns(memory_slice(p, sizeof(int16_t) * MLKEM_N))
+  ensures(array_abs_bound(p, 0, MLKEM_N, MLKEM_Q))
+);
 
 #define mlk_poly_mulcache_compute_asm MLK_NAMESPACE(poly_mulcache_compute_asm)
 void mlk_poly_mulcache_compute_asm(int16_t *, const int16_t *, const int16_t *,
                                    const int16_t *);
 
 #define mlk_poly_tobytes_asm MLK_NAMESPACE(poly_tobytes_asm)
-void mlk_poly_tobytes_asm(uint8_t *r, const int16_t *a);
+void mlk_poly_tobytes_asm(uint8_t *r, const int16_t *a)
+/* This must be kept in sync with the HOL-Light specification
+ * in proofs/hol_light/arm/proofs/mlkem_poly_tobytes.ml */
+__contract__(
+  requires(memory_no_alias(r, MLKEM_POLYBYTES))
+  requires(memory_no_alias(a, sizeof(int16_t) * MLKEM_N))
+  requires(array_bound(a, 0, MLKEM_N, 0, MLKEM_Q))
```

Comment:
This is not assumed in the HOL-Light spec, actually

Reply:
Hmm, I'm not sure I agree with this. Yes, technically the HOL-Light spec works for 12-bit integers.
But this function should _never_ be called on anything outside of [0,q). Otherwise you won't match the testvectors and fail the key validation. 

I've changed it to `MLKEM_UINT12_LIMIT` for now.

Reply:
Note that the wrappers around this function _do_ have the bounds condition, so you don't loose anything by sticking exactly to the HOL-Light spec. I think it is confusing if we start deviating between the HOL-Light spec and the immediate reflection in CBMC.

Reply:
Okay.